### ISSUE TITLE
Pyright does not recognise that an empty TypedDict is falsy

### ISSUE THREAD

**Issue by strutt** [microsoft/pyright#9370]
**Describe the bug**
Pyright does not recognise that an empty TypedDict is falsy. In the `dict_factory_caller` method below pyright fails to narrow the type of `x` to `NotEmptyDict` inside the second `if` statement.

**Code or Screenshots**

```python
from typing import Literal, TypedDict, overload, reveal_type


class NotEmptyDict(TypedDict):
    item: str


class EmptyDict(TypedDict):
    pass


@overload
def dict_factory(_type: Literal["empty"]) -> EmptyDict:
    ...


@overload
def dict_factory(_type: Literal["notEmpty"]) -> NotEmptyDict:
    ...


def dict_factory(
    _type: Literal["empty"] | Literal["notEmpty"],
) -> NotEmptyDict | EmptyDict:
    if _type == "empty":
        return {}
    return {"item": "value"}


def dict_factory_caller(_type: Literal["empty", "notEmpty"]) -> None | str:
    x = dict_factory(_type)

    if "item" in x:
        reveal_type(x)  # information: Type of "x" is "NotEmptyDict"

    if x:
        reveal_type(x)  # information: Type of "x" is "NotEmptyDict | EmptyDict"
        return x["item"]  # error: Could not access item in TypedDict
    return None
```

**VS Code extension or command-line**
command line: pyright 1.1.387

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

TypedDicts are structural types. A value that conforms to a TypedDict can have additional keys that are not specified by the TypedDict. Such values would not be falsy. It would therefore be incorrect (and unsafe) for a type checker to narrow the type in this case.

I recommend using a different check in this case if you want your code to be type safe.
